proposals/lip-0049.md

```diff
@@ -223,135 +223,171 @@ When processing a cross-chain message `ccm`, the functions below may be called b
-            ccm.moduleID,
-            ccm.crossChainCommandID,
-            ccm.sendingChainID,
-            0,
-            CCM_STATUS_MODULE_NOT_SUPPORTED,
-            ccm.params
+    # If the status is not ok and the fee is sufficient, we send back an error message.
+    elif ccm.status == CCM_STATUS_OK and ccm.fee >= MIN_RETURN_FEE*size(ccm):
+        if moduleID is not supported:
+            errorStatus = CCM_STATUS_MODULE_NOT_SUPPORTED
+            eventStatus = CCM_PROCESSED_STATUS_MODULE_NOT_SUPPORTED
+        elif crossChainCommandID is not supported:
+            errorStatus = CCM_STATUS_CROSS_CHAIN_COMMAND_NOT_SUPPORTED
+            eventStatus = CCM_PROCESSED_STATUS_COMMAND_NOT_SUPPORTED
+
+        emitEvent(
+            moduleID = MODULE_ID_INTEROPERABILITY,
+            typeID = TYPE_ID_CCM_PROCESSED,
+            data = { "ccmID": ccmID, "status": eventStatus},
+            topics = [ccm.sendingChainID, ccm.receivingChainID]
         )
 
-    elif crossChainCommandID is not supported:
-        sendInternal(
+        send(
             EMPTY_FEE_ADDRESS,
             ccm.moduleID,
             ccm.crossChainCommandID,
             ccm.sendingChainID,
             0,
-            CCM_STATUS_CROSS_CHAIN_COMMAND_NOT_SUPPORTED,
+            errorStatus,
             ccm.params
         )
+    # Else, we discard the message.
+    else:
+        emitEvent(
+            moduleID = MODULE_ID_INTEROPERABILITY,
+            typeID = TYPE_ID_CCM_PROCESSED,
+            data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_COMMAND_DISCARDED},
+            topics = [ccm.sendingChainID, ccm.receivingChainID]
+        )
 ```
 
 #### forward
 
-The `forward` function forwards a CCM to the specified receiving chain. This function is only called on the mainchain when processing a CCU.
+The `forward` function forwards a CCM to the specified receiving chain. This function is only called on the mainchain when processing a CCU. If an error occurs, the CCM is bounced back to the sending chain or discarded.
 
 ```python
-# The ccu including this message is passed as context to this function
-forward(ccu, ccm):
-    if (chainAccount(ccm.receivingChainID) exists
-        and chainAccount(ccm.receivingChainID).status == CHAIN_ACTIVE
-        and isLive(ccm.receivingChainID):
-        # The Token module is called to transfer the fee from sending to receiving chain
-        token.forwardMessageFee(ccm)
-        if the previous call fails or returns False:
-            return
+def forward(ccu: CCU, ccm: CCM) -> None:
+    # Calculate CCM ID, used later in events.
+    ccmID = SHA-256(encode(crossChainMessageSchema, ccm))
+    
+    # Chain with status REGISTERED do not pass this check.
+    if isLive(ccm.receivingChainID) and chainAccount(ccm.receivingChainID).status == CHAIN_STATUS_ACTIVE:
+        # Transfer the fee from escrow to escrow, call Token LIP
+        Token.transferBetweenEscrows(ccm.sendingChainID, ccm.receivingChainID, feeTokenID, fee)
```

Comment:
also here fee should be `fee = ccm.fee` and `feeTokenID` recovered from the channel.

Reply:
Is this the only place where the Interoperability module explicitly calls the Token module? Would it make sense to have a hook for ccm forwarding instead?

Reply:
We discussed this in the past and we said the less the hooks the better. Having a hook for forwarding is also not so nice since it would be used only on mainchain (and only for the token module most likely)

Reply:
I created a discussion on Slack for this as I would see additional benefits, e.g., being able to remove the cross-chain forward message.

Reply:
As we concluded to add the hook for forwarding, the comment can be resolved from my side.